File: cppsocket/sender.cpp
#include <chrono>
#include <iostream>
#include <fstream>
#include <cstring>
#include <arpa/inet.h>
#include <unistd.h>

#define PORT 5000
#define BUFFER_SIZE 4096

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: ./sender <Receiver IP> <File path>\n";
        return 1;
    }

    const char* server_ip = argv[1];
    const char* filepath = argv[2];

    // Open file
    std::ifstream input_file(filepath, std::ios::binary);
    if (!input_file) {
        std::cerr << "Cannot open file.\n";
        return 1;
    }

    // Create socket
    int sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) {
        perror("Socket creation failed");
        return 1;
    }

    sockaddr_in server_addr;
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(PORT);

    if (inet_pton(AF_INET, server_ip, &server_addr.sin_addr) <= 0) {
        perror("Invalid address");
        return 1;
    }

    // Connect to server
    if (connect(sock, (sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        perror("Connection failed");
        return 1;
    }
    std::cout << "Connected to receiver.\n";

    // Send filename size and filename
    std::string filename = std::string(filepath).substr(std::string(filepath).find_last_of("/\\") + 1);
    uint32_t filename_size = htonl(filename.size());
    send(sock, &filename_size, sizeof(filename_size), 0);
    send(sock, filename.c_str(), filename.size(), 0);

    // Send file data
    char buffer[BUFFER_SIZE];
    auto start_time = std::chrono::high_resolution_clock::now();
    while (input_file.read(buffer, BUFFER_SIZE) || input_file.gcount() > 0) {
        send(sock, buffer, input_file.gcount(), 0);
    }
    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::seconds>(end_time - start_time).count();

    std::cout << "File sent successfully.\n";

    // Close connections
    input_file.close();
    close(sock);

    return 0;
}

File: pythonsocket/sender.py
import os
import time
import socket

FILENAME = "data.bin"
HOST = "192.168.0.120"
PORT = 5000
BUFFER_SIZE = 16384

if not os.path.exists(FILENAME):
    print(f"File {FILENAME} does not exist")
    exit(1)

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.connect((HOST, PORT))
print(f"Connected to {HOST}:{PORT}")

filename_encoded = FILENAME.encode("utf-8")
filename_length = len(filename_encoded)
s.send(filename_length.to_bytes(4, "big"))
s.send(filename_encoded)

print(f"Sending file: {FILENAME}...")
t0 = time.time()
fp = open(FILENAME, "rb")
nbytes = 0
while True:
    data = fp.read(BUFFER_SIZE)
    nbytes += len(data)
    if not data:
        break
    s.send(data)
    print(f"Sent {nbytes / 1e9:.2f} GB [{nbytes / (time.time() - t0) / 1e6:.1f} MB/s]   ", end="\r")
print(f"Sent {nbytes / 1e9:.2f} GB [{nbytes / (time.time() - t0) / 1e6:.1f} MB/s]")
fp.close()
t1 = time.time()
print(f"File {FILENAME} sent successfully")
print(f"File size: {nbytes / 1e9:.3f} GB")
print(f"Time taken: {t1 - t0:.2f} seconds")
print(f"Average speed: {(nbytes / (t1 - t0)) / 1e6:.3f} MB/s")

s.close()
